src/Rheologies/elasto_visco_plastic_rheology.jl

```diff
@@ -0,0 +1,307 @@
+                               ice_compaction_hardening = 20, 
+                               yield_curve_eccentricity = 2, 
+                               Δ_min = 1e-10,
+                               min_substeps = 30,
+                               max_substeps = 500)
+
+Constructs an `ElastoViscoPlasticRheology` object representing a "modified" elasto-visco-plastic
+rheology for slab sea ice dynamics that follows the implementation of Kimmritz et al (2016).
+The stress tensor is computed following the constitutive relation:
+```math
+σᵢⱼ = 2η ϵ̇ᵢⱼ + [(ζ - η) * (ϵ̇₁₁ + ϵ̇₂₂) - P / 2] δᵢⱼ
+```
+where ``ϵ̇ᵢⱼ`` are the strain rates, ``η`` is the shear viscosity, ``ζ`` is the bulk viscosity,
+and ``P`` is the ice strength (acting as the isotropic part of the stress tensor)
+parameterized as ``P★ h exp( - C ⋅ ( 1 - ℵ ))`` where ``P★`` is the `ice_compressive_strength`, 
+``C`` is the `ice_compaction_hardening`, ``h`` is the ice thickness, and ``ℵ`` is the ice concentration.
+
+The stresses are substepped using a dynamic substepping coefficient ``α`` that is
+spatially varying and computed dynamically as in Kimmritz et al (2016)
+In particular: α = sqrt(γ²) 
+where γ² = ζ * π² * (Δt / mᵢ) / Az is a stability parameter with ``Az`` is the area of the grid cell, 
+``mᵢ`` the ice mass, and ``Δt`` the time step.
+
+The stresses are substepped with:
+```math
+σᵢⱼᵖ⁺¹ = σᵢⱼᵖ + (σᵢⱼᵖ⁺¹ - σᵢⱼᵖ) / α
+```
+
+This formulation allows fast convergence in regions where α is small. Regions where
+α is large correspond to regions where the ice is more solid and the convergence is slower.
+
+The number of substeps is then bounded by `min_substeps` and `max_substeps`.
+
+Arguments
+=========
+    
+- `grid`: the `SlabSeaIceModel` grid
+
+Keyword Arguments
+=================
+    
+- `ice_compressive_strength`: parameter expressing compressive strength (in Nm²). Default `27500`.
+- `ice_compaction_hardening`: exponent coefficient for compaction hardening. Default `20`.
+- `yield_curve_eccentricity`: eccentricity of the elliptic yield curve. Default `2`.
+- `Δ_min`: Minimum value for the visco-plastic parameter. Limits the maximum viscosity of the ice, 
+           transitioning the ice from a plastic to a viscous behaviour. Default value is `1e-10`.
+- `min_substeps`: Minimum number of substeps expressed as the dynamic coefficient. Default value is `30`.
+- `max_substeps`: Maximum number of substeps expressed as the dynamic coefficient. Default value is `500`.
+"""
+function ElastoViscoPlasticRheology(FT::DataType = Float64; 
+                                    ice_compressive_strength = 27500, 
+                                    ice_compaction_hardening = 20, 
+                                    yield_curve_eccentricity = 2, 
+                                    minimum_plastic_stress = 2e-9,
+                                    min_substeps = 50,
+                                    max_substeps = 1000)
```

Comment:
is it actually more performant to use variable steps on GPU, vs simply always going to max substeps?

Reply:
We should probably call this in another way, these are fake substeps that help convergence, the substepping is fixed to the substeps passed to the `SpliExplicitSolver`